### PR TITLE
feat: CameraUtil and image <-> byte[]

### LINKED ISSUE #52: hacktoberfest java | camera stream byte[] utility
## Premise

One of the things the atPlatform specializes in is sending end-to-end encrypted binary data. A huge application would be to send a camera stream through atSigns.

## What to do
Create a utility class to fetch the camera stream from a USB camera (or a camera from your choice) and convert it into a Map of images.
Perhaps start with taking a screenshot of the camera stream and converting that into a byte[] array.

## Support
In case you need any kind of help (such as getting your environment setup or asking questions about our tech) don’t be afraid to ask in our [discord](https://discord.atsign.com). There is a team full of developers ready to jump in and help on whatever your issue may be.

### LINKED ISSUE #54: hacktoberfest java | image <-> byte[] utility
## Premise

The atPlatform is perfect for sending data and files peer-to-peer. The atPlatform ensures that the data sent is only readable by the end user.

## What needs to be done

- Create 2 utility methods (one for converting an image to an array of bytes (byte[]) and another for converting the byte[] image to an image object)
- You can do this in anyway you like (does not have to be a method, could be an entire class! or multiple classes)
- Even codeblocks are appreciated

## Support
In case you need any kind of help (such as getting your environment setup or asking questions about our tech) don’t be afraid to ask in our [discord](https://discord.atsign.com). There is a team full of developers ready to jump in and help on whatever your issue may be.

### PR DESCRIPTION
**Pull request by kaggrwal**
**- Description for the changelog**
Two Utility classes are added
1. CameraUtil 
 This Utility class can be used to capture a stream from the camera and get the result in list of byte[] OR capture a single image from camera into byte[]. Camera name can be provided to chose which camera would be used to capture.

2. ImageUtil
This utility class can be used to convert buffered image to byte[] and vice versa, it can also be used to save image to memory
Image format and file location can be provided, default format is 'jpg' and default location would be classpath

How to use: Check the comments in both classes, examples have been provided.
closes #52
closes #54

### CODE REVIEW

**Comment by kaggrwal** [comment]
@JeremyTubongbanua  please review and let me know if it requires any improvements. This request fixes #52 , #54

File: at_client/src/main/java/org/atsign/client/util/CameraUtil.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ *      //example for camera Name: "Webcam HD Webcam: HD Webcam /dev/video0"
+ *      //Check you system drivers to identify camera name or use getAllCams method here
+ *
+ */
+
+public class CameraUtil{
+
+    public static final String START = "START";
+    public static final String STOP = "STOP";
+    public static final String EXIT = "EXIT";
```
Comment (mutairibassam) [inline at_client/src/main/java/org/atsign/client/util/CameraUtil.java]: I would recommend to use `enum types` instead.

**Comment by JeremyTubongbanua** [comment]
@kaggrwal What kind of camera does this work for?

**Comment by kaggrwal** [comment]
> @kaggrwal What kind of camera does this work for?

@JeremyTubongbanua, This would work with any camera that is attached to the system (any inbuilt webcam or webcam attached through usb). 

I have provided the option to provide camera name, otherwise code would pick the default camera provided by OS. 
`//getCameraStream(String cameraName)`

One can also get the names of all cameras that are installed in the system 
`//List<String> getAllCams`


Additionally, we can extend this framework to use IP cameras also.

File: at_client/src/main/java/org/atsign/client/util/CameraUtil.java (hunk 1 above)
Comment (kaggrwal) [reply at_client/src/main/java/org/atsign/client/util/CameraUtil.java]: This was not being used now, i have deleted them.

**Comment by JeremyTubongbanua** [comment]
That's amazing @kaggrwal :)

Do you have any idea why I am getting this? I am running this on a Macbook Air with the M2 chip (running macOS). 

<details>
<summary>
Error message
</summary>
<pre>
<code>
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
java.lang.RuntimeException: Failed to initialize BridJ (java.lang.UnsatisfiedLinkError: Can't load library: /var/folders/lv/j4q1thzj5vg9zc50z0k11brc0000gn/T/BridJExtractedLibraries13213607309581796019/libbridj.dylib)
        at org.bridj.Platform.initLibrary(Platform.java:444)
        at org.bridj.Platform.<clinit>(Platform.java:228)
        at com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.<clinit>(OpenIMAJGrabber.java:59)
        at com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$WebcamNewGrabberTask.handle(WebcamDefaultDriver.java:55)
        at com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.run(WebcamProcessor.java:81)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.UnsatisfiedLinkError: Can't load library: /var/folders/lv/j4q1thzj5vg9zc50z0k11brc0000gn/T/BridJExtractedLibraries13213607309581796019/libbridj.dylib
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2398)
        at java.base/java.lang.Runtime.load0(Runtime.java:785)
        at java.base/java.lang.System.load(System.java:1979)
        at org.bridj.Platform.initLibrary(Platform.java:420)
        ... 7 more
Exception in thread "main" com.github.sarxos.webcam.WebcamException: java.util.concurrent.ExecutionException: com.github.sarxos.webcam.WebcamException: Cannot execute task
        at com.github.sarxos.webcam.WebcamDiscoveryService.getWebcams(WebcamDiscoveryService.java:124)
        at com.github.sarxos.webcam.Webcam.getWebcams(Webcam.java:893)
        at com.github.sarxos.webcam.Webcam.getWebcams(Webcam.java:866)
        at com.github.sarxos.webcam.Webcam.getWebcams(Webcam.java:845)
        at org.atsign.client.util.CameraUtil.getAllCams(CameraUtil.java:73)
        at org.atsign.common.CameraUtilTest.main(CameraUtilTest.java:11)
Caused by: java.util.concurrent.ExecutionException: com.github.sarxos.webcam.WebcamException: Cannot execute task
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at com.github.sarxos.webcam.WebcamDiscoveryService.getWebcams(WebcamDiscoveryService.java:116)
        ... 5 more
Caused by: com.github.sarxos.webcam.WebcamException: Cannot execute task
        at com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.process(WebcamProcessor.java:72)
        at com.github.sarxos.webcam.WebcamProcessor.process(WebcamProcessor.java:140)
        at com.github.sarxos.webcam.WebcamTask.process(WebcamTask.java:46)
        at com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$WebcamNewGrabberTask.newGrabber(WebcamDefaultDriver.java:45)
        at com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver.getDevices(WebcamDefaultDriver.java:117)
        at com.github.sarxos.webcam.WebcamDiscoveryService$WebcamsDiscovery.call(WebcamDiscoveryService.java:36)
        at com.github.sarxos.webcam.WebcamDiscoveryService$WebcamsDiscovery.call(WebcamDiscoveryService.java:26)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.UnsatisfiedLinkError: 'int org.bridj.Platform.sizeOf_ptrdiff_t()'
        at org.bridj.Platform.sizeOf_ptrdiff_t(Native Method)
        at org.bridj.Platform.<clinit>(Platform.java:232)
        at com.github.sarxos.webcam.ds.buildin.natives.OpenIMAJGrabber.<clinit>(OpenIMAJGrabber.java:59)
        at com.github.sarxos.webcam.ds.buildin.WebcamDefaultDriver$WebcamNewGrabberTask.handle(WebcamDefaultDriver.java:55)
        at com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.run(WebcamProcessor.java:81)
        ... 3 more
</code>
</pre>
</details>
<details>
<summary>
Main method
</summary>
<pre>
<code>
package org.atsign.common;

import org.atsign.client.util.CameraUtil;
import java.awt.image.BufferedImage;
import java.util.List;

public class CameraUtilTest {

    public static void main(String[] args) {
        // BufferedImage image = CameraUtil.getSingleImage();
        List<String> cams = CameraUtil.getAllCams();
        System.out.println(cams);

    }

}
</code>
</pre>
</details>

**Comment by kaggrwal** [comment]
> libbridj.dylib

Hi @JeremyTubongbanua , Due to permissions' issue on macOS, we need to install additional dependencies which are more related to macOS.

Please see this link:https://github.com/sarxos/webcam-capture/issues/757#issuecomment-849632328

**Comment by kaggrwal** [comment]
@JeremyTubongbanua Did you get time to look at this ?

**Comment by JeremyTubongbanua** [comment]
Hey @kaggrwal I get this now when running it on a Windows PC:

```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

**Comment by JeremyTubongbanua** [comment]
Nvm- It works! @kaggrwal :)

**Review by JeremyTubongbanua — APPROVED** [review]
Awesome work @kaggrwal

**Comment by JeremyTubongbanua** [comment]
@kaggrwal 
Hey! Thank you for your contribution. Your PR was accepted, merged, and it closed an issue. You could be eligible for a prize. 

To be eligible, please send an email to [hacktoberfest@atsign.com](mailto:hacktoberfest@atsign.com) with a screenshot of the PR approval email. This is to prove that your email corresponds to the correct GitHub account. Thank you!

When it comes time to distribute the prizes, we will look at all the emails we've received from that email so don't forget to do this!

**Pull request merged**